CorsixTH/Lua/humanoid_actions/knock_door.lua

```diff
@@ -26,41 +26,70 @@ local KnockDoorAction = _G["KnockDoorAction"]
 --! Constructor for knocking on the door action.
 --!param door (Object) Door to knock on.
 --!param direction (string) Direction of facing.
-function KnockDoorAction:KnockDoorAction(door, direction)
+function KnockDoorAction:KnockDoorAction(humanoid, door, direction)
+  assert(class.is(humanoid, Humanoid), "Invalid value for parameter 'humanoid'")
   assert(class.is(door, Door), "Invalid value for parameter 'door'")
   assert(direction == "north" or direction == "south" or
       direction == "east" or direction == "west",
       "Invalid value for parameter 'direction'")
 
   self:HumanoidAction("knock_door")
+  self.humanoid = humanoid
   self.door = door -- Door to knock on.
   self.direction = direction -- Direction of facing.
+  self.state = nil -- Next step in the animation.
+end
+
+function KnockDoorAction:start()
+  self.state = 0
+  self:update()
+end
+
+
+local function action_knock_door_start(action, humanoid)
+  assert(humanoid == action.humanoid)
+  action:start()
 end
 
 local action_knock_door_tick = permanent"action_knock_door_tick"( function(humanoid)
-  local door = humanoid.user_of
-  door:removeUser(humanoid)
-  humanoid.user_of = nil
-  door:getRoom():tryAdvanceQueue()
-  humanoid:finishAction()
+  local action = humanoid:getCurrentAction()
+  action:update()
 end)
 
-local function action_knock_door_start(action, humanoid)
-  local direction = action.direction
-  local anims = humanoid.door_anims
-  local door = action.door
-  action.must_happen = true
-  local anim = anims.knock_north
-  local flag_mirror = (direction == "west" or direction == "south") and 1 or 0
-  if direction == "east" or direction == "south" then
-    anim = anims.knock_east
+
+function KnockDoorAction:update()
+  if self.state == 0 then
+    local direction = self.direction
+    local anims = self.humanoid.door_anims
+    local door = self.door
+    self.must_happen = true
+    local anim = anims.knock_north
+    local flag_mirror = (direction == "west" or direction == "south") and 1 or 0
+    if direction == "east" or direction == "south" then
+      anim = anims.knock_east
+    end
+    self.humanoid:setAnimation(anim, flag_mirror)
```

Comment:
Save compat issues I think. self.humanoid wasn't defined in the old action.

Reply:
Ah yes indeed, look for `KnockDoorAction` instances in all action queues of humanoids, and add `action.humanoid`

Reply:
rebased the branch and fixed the issue by adding a 3rd commit